rust-webvr/src/api/glwindow/display.rs

```diff
@@ -0,0 +1,261 @@
+            right_degrees: fov_right,
+            up_degrees: fov_up,
+        };
+
+        let left_eye_parameters = VREyeParameters {
+            offset: [-EYE_DISTANCE / 2.0, 0.0, 0.0],
+            render_width: self.size.width as u32 / 2,
+            render_height: self.size.height as u32,
+            field_of_view: field_of_view,
+        };
+
+        let right_eye_parameters = VREyeParameters {
+            offset: [EYE_DISTANCE / 2.0, 0.0, 0.0],
+            ..left_eye_parameters.clone()
+        };
+
+        VRDisplayData {
+            display_id: self.id,
+            display_name: self.name.clone(),
+            connected: true,
+            capabilities: capabilities,
+            stage_parameters: None,
+            left_eye_parameters: left_eye_parameters,
+            right_eye_parameters: right_eye_parameters,
+        }
+    }
+
+    fn immediate_frame_data(&self, near: f64, far: f64) -> VRFrameData {
+        GlWindowVRDisplay::frame_data(0.0, self.size, near, far)
+    }
+
+    fn synced_frame_data(&self, near: f64, far: f64) -> VRFrameData {
+        self.immediate_frame_data(near, far)
+    }
+
+    fn reset_pose(&mut self) {}
+
+    fn sync_poses(&mut self) {}
+
+    fn future_frame_data(&mut self, near: f64, far: f64) -> VRFutureFrameData {
+        let (resolver, result) = VRFutureFrameData::blocked();
+        let _ = self.sender.send(GlWindowVRMessage::StartFrame(near, far, resolver));
+        result
+    }
+
+    fn bind_framebuffer(&mut self, _eye_index: u32) {}
+
+    fn get_framebuffers(&self) -> Vec<VRFramebuffer> {
+        let left_viewport = VRViewport {
+            x: 0,
+            y: 0,
+            width: (self.size.width as i32) / 2,
+            height: self.size.height as i32,
+        };
+
+        let right_viewport = VRViewport {
+            x: self.size.width as i32 - left_viewport.width,
+            ..left_viewport
+        };
+
+        vec![
+            VRFramebuffer {
+                eye_index: 0,
+                attributes: VRFramebufferAttributes::default(),
+                viewport: left_viewport,
+            },
+            VRFramebuffer {
+                eye_index: 1,
+                attributes: VRFramebufferAttributes::default(),
+                viewport: right_viewport,
+            },
+        ]
+    }
+
+    fn render_layer(&mut self, _layer: &VRLayer) {
+        unreachable!()
+    }
+
+    fn submit_frame(&mut self) {
+        unreachable!()
+    }
+
+    fn submit_layer(&mut self, gl: &Gl, layer: &VRLayer) {
+        // TODO: this assumes that the current GL framebuffer contains the texture
+        // TODO: what to do if the layer has no texture_size?
+        if let Some((width, height)) = layer.texture_size {
+            let num_bytes = (width as usize) * (height as usize) * 4;
+            let mut buffer = self.pool.remove().unwrap_or_else(Vec::new);
+            buffer.resize(num_bytes, 0);
+            gl.read_pixels_into_buffer(
```

Comment:
Can we share the data in a more performant way?

Reply:
We'd need to have GL context sharing, which got added to glutin after this PR was submitted (https://github.com/tomaka/glutin/issues/899#issuecomment-467239837). Context sharing is still not implemented on macOS. This implementation makes all platforms use a slow path, to get other platforms onto the fast path, we'd have to recreate the interface that webrender has for images that are either implemented using shared textures or using readback. I'm not sure how to do that without making `rust-webvr` depend on `webrender_api`.

Reply:
The spec issue for this is https://github.com/immersive-web/webxr/issues/528. The XR spec pretty much requires textures to go via main memory in the case that the browser doesn't share GL textures with the VR display.